src/card-draw.ts

```diff
@@ -307,62 +307,90 @@ export function draw(gameData: GameData, configData: ConfigState): Drawing {
     }
   }
 
+  // OK, setup work is done, here's whre we actually draw the cards!
+
+  let redraw = false;
   const drawnCharts: DrawnChart[] = [];
-  /**
-   * Record of how many songs of each bucket index have been drawn so far
-   */
-  const difficultyCounts = new CountingSet<number>();
 
-  // OK, setup work is done, here's whre we actually draw the cards!
-  while (drawnCharts.length < numChartsToRandom) {
-    if (bucketDistribution.length === 0) {
-      // no more songs available to pick in the requested range
-      // will be returning fewer than requested number of charts
-      break;
+  do {
+    /**
+     * Record of how many songs of each bucket index have been drawn so far
+     */
+    const difficultyCounts = new CountingSet<number>();
+
+    // make a copy of valid charts here in the loop so we
+    // can mutate it later during the draw process, but
+    // start with a fresh copy each full draw attempt
+    const localValidCharts = new DefaultingMap<number, EligibleChart[]>(
+      () => [],
+    );
+    for (const [bucketIdx, charts] of validCharts) {
+      // make a clone of each inner array, too
+      localValidCharts.set(bucketIdx, charts.slice());
     }
```

Comment:
And I'm pretty sure this will work to correctly "reset the deck" before each pass through the outer loop